working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,411 @@
+/// only that *if* it is called with a finalization token as argument,
+/// at least one object registered in the registry with that finalization token
+/// is no longer accessible to the program.
+///
+/// If the same object is registered in multiple finalization registries,
+/// or registered multiple times in a single registry,
+/// and the object becomes inaccessible to the program,
+/// then any number of those registrations may trigger their associated
+/// callback. It will not necessarily be all or none of them.
+///
+/// Finalization callbacks will happen as *events*, not during execution of
+/// other code and not as a microtask, but as high-level events similar to
+/// timer events.
+abstract class FinalizationRegistry<FT> {
+  /// Creates a finalization registry with the given finalization callback.
+  external factory FinalizationRegistry(
+    void Function(FT finalizationToken) callback);
+
+  /// Registers [value] for a finalization callback.
+  ///
+  /// When [value] is no longer accessible to the program,
+  /// the registry *may* call its callback function with [finalizationToken]
+  /// as argument.
+  ///
+  /// The [value] and [unregisterToken] arguments do not count towards those
+  /// objects being accessible to the program. Both must be objects supported
+  /// as an [Expando] key.
+  ///
+  /// Multiple objects may be registered with the same finalization token,
+  /// and the same object may be registered multiple times with different,
+  /// or the same, finalization token.
+  ///
+  /// The callback may be called at most once per registration, and not
+  /// for registrations which have been unregistered since they were registered.
+  void register(Object value, FT finalizationToken, {Object? unregisterToken});
+
+  /// Unregisters any finalization callbacks registered with [unregisterToken]
+  /// as unregister-token.
+  ///
+  /// After unregistering, those callbacks will not happen even if the
+  /// registered object becomes inaccessible.
+  void unregister(Object? unregisterToken);
+}
+
+/// A weak reference to another object.
+///
+/// A _weak_ reference to the [target] object which may be cleared
+/// (set to reference `null` instead) at any time
+/// where there is no other ways for the program to access the target object.
+///
+/// _The referenced object may be garbage collected when the only reachable
+/// references to it are weak._
+///
+/// Not all objects are supported as targets for weak references. [WeakRef]
+/// constructor and [WeakRef.target] will reject any object that is not
+/// supported as an [Expando] key.
+abstract class WeakRef {
+  /// Create a [WeakRef] pointing to the given [target], which must be
+  /// an object supported as an [Expando] key.
+  external factory WeakRef(Object? target);
+
+  /// The current object weakly referenced by [this]. Is either [null] or
+  /// an object supported as an [Expando] key.
+  abstract Object? target;
```

Comment:
Is there a reason this can't be a generic?

Reply:
I don't see any issue, other than the extra memory overhead of storing the type.

The design question is really whether this is a low-level or high-level primitive. 

A low-level primitive is fine without fancy generic types, you're supposed to build that, and other abstractions, on top of it. Doing nothing by itself ensures maximal flexibility.

A more high-level primitive should fit directly into application code, and should definitely be typed.

So, what are we aiming for here? Directly usable by all our developers, or a building block only intended for those doing specialized libraries on top?

Reply:
For @jacob314 use case I don't think we'd need to typecast anyway for this to be useful. However, it seems like most use cases would want to typecast. I'm under the impression that would be cheaper if this was just generic, but maybe not.

Reply:
We should add a generic. Adding a generic shouldn't have much overhead and will make the API easier to use. As some prior art, the similar API for typescript has a generic (of course zero cost as the generic is erased).
https://github.com/microsoft/TypeScript/blob/main/src/lib/es2021.weakref.d.ts